### ISSUE TITLE
eliminate floating 

### ISSUE THREAD

**Issue by Bin-ze** [Totoro97/f2-nerf#61]
Thanks for your new perspective and great work! I tried to use f2-nerf for large scene reconstruction and got good results.

But there will be floating on some paths, which affects the final performance.

Do you have any ideas for eliminating floaters? I tried parallax regularized loss, but it makes the geometry worse.

Can you give me some advice, dear author!

**Comment by Totoro97** [comment]
Hi, I add a gradient scaling strategy to penalize gradients of near samplings, which may help reduce the floaters. You can try the latest commit using the config `wanjinyou.yaml` or `wanjinyou_big.yaml`. Look forward to your feedback😃

**Comment by Bin-ze** [comment]
Is the cause of floating related to the algorithm's octree-based space division strategy? I observed the result, the value of the estimated depth of the position where the floating object is generated is invalid, I am a bit confused why this is the case, can you give some advice?
![image](https://github.com/Totoro97/f2-nerf/assets/72087870/ada3ac2a-d5c7-4293-8539-3781bf3ace8b)

**Comment by Bin-ze** [comment]
I use  gradient scaling strategy and it help me reduce the floaters, but new problem arises，The depth map becomes discontinuous, and there will be holes on the border close to the camera. This should be caused by the gradient clipping strategy. How to adjust it can alleviate such problems？
![image](https://github.com/Totoro97/f2-nerf/assets/72087870/8ae50d94-296a-41bf-894c-ff6d6e561d38)

**Comment by Totoro97** [comment]
Hi, I think this effect is in expectation. The depth values are difficult to correctly be estimated for the textureless property of the white walls. Maybe add some depth constraints using some pre-trained depth estimation networks. I feel sorry that at this stage I am not able to think of a simple quick solution to alleviate such problems.

**Comment by Bin-ze** [comment]
Thanks for the reply, but this behavior occurs after adding gradient scaling, before that, the depth of the wall looks fine. The previous problem was that in a certain section of the rendering path, there would be some floating objects, as shown in：
![image](https://github.com/Totoro97/f2-nerf/assets/72087870/afff84f8-e198-43da-8b77-f93638f74510)
I added gradient scaling, the previous floating objects disappeared, but the wall close to the camera became fragmented, how to control the value of gradient scaling becomes close to 1 as it moves away from the camera position, maybe it can be adjusted manually to reduce the impact on the wall Impact

**Comment by Bin-ze** [comment]
How does gradient scaling regularize samples close to the camera? I checked the implementation:

      __global__ void GradientScalingBackwardKernel(int n_rays, int c, float progress, float* rand_val, int* idx_start_end, float* out_vals) {
        int idx = LINEAR_IDX();
        if (idx >= n_rays) return;
        int idx_start = idx_start_end[idx * 2];
        int idx_end   = idx_start_end[idx * 2 + 1];
        for (int i = 0; i + idx_start < idx_end; i++) {
          float a = (float(i) + .5f) / float(idx_end - idx_start);
          float cur_scale = progress + (1.f - progress) * a * a;
          for (int j = 0; j < c; j++) {
            out_vals[(i + idx_start) * c + j] *= cur_scale;
          }
        }
      }

Calculate the scaling factor here：
> float cur_scale = progress + (1.f - progress) * a * a;

It gradually decreases to 1 as a increases, and **a** represents the normalized distance of the sampling interval on the light.
 
Does it mean that this strategy will give a larger gradient to the sampling points close to the camera, so that the model pays more attention to the part？
